### PR TITLE
Introduce class keyword for refs, allow type refs without ()

### LINKED ISSUE #1508: Allow importing of types without the use of () if no constructors are required
Currently the behaviour of an import for a proper name is dictated by syntax, but that's basically due to laziness on my part. Also, importing `X()` is a bit pointless, especially in the case of synonyms where `()` can only ever be empty.

### PR DESCRIPTION
**Pull request by garyb**
Resolves #1508

Or will do when I fix whatever stupid mistake I've made that is breaking the tests.

### CODE REVIEW

**Comment by garyb** [comment]
There we go. Pesky `Positioned` constructors.

**Comment by michaelficarra** [comment]
You've not added any new tests.

**Comment by garyb** [comment]
True. We can't test for warnings yet, so I can't add tests for that, but I'll add a test for the `class` syntax and type without `()`.

**Comment by garyb** [comment]
Added tests, thanks for the reminder @michaelficarra.

**Comment by michaelficarra** [comment]
How about in explicit export list?

**Comment by garyb** [comment]
Good thinking, added those too.

File: psci/PSCi.hs [hunk 1]
```diff
@@ -336,6 +324,7 @@ handleShowImportedModules = do
   showRef (P.TypeRef pn dctors) = show pn ++ "(" ++ maybe ".." (commaList . map N.runProperName) dctors ++ ")"
   showRef (P.ValueRef ident) = show ident
```
Comment (paf31) [inline psci/PSCi.hs]: Should these really be using `Show`?

File: src/Language/PureScript/Errors.hs [hunk 2]
```diff
@@ -781,6 +771,14 @@ prettyPrintSingleError full level e = do
       paras [ line $ "The import of type " ++ runModuleName name ++ " as " ++ runModuleName qualName ++ " uses the deprecated 'import qualified' syntax."
             , line $ "This syntax form will be removed in PureScript 0.9." ]
 
+    renderSimpleErrorMessage (DeprecatedClassImport mn name) =
+      paras [ line $ "The import of class " ++ runProperName name ++ " from " ++ runModuleName mn ++ " uses the deprecated syntax that omits the class keyword."
```
Comment (paf31) [inline src/Language/PureScript/Errors.hs]: Does it make sense to show the new syntax in full for the import in question?

File: psci/PSCi.hs (hunk 1 above)
Comment (garyb) [inline psci/PSCi.hs]: I don't know, but I suspect not!

File: src/Language/PureScript/Errors.hs (hunk 2 above)
Comment (garyb) [inline src/Language/PureScript/Errors.hs]: We'd have to reconstruct the whole import/export declaration to be able to do that in a useful manner I think - what can we show them apart from `class SomeProperName`?

**Comment by garyb** [comment]
Rebased and fixed that `show` usage in `psci`.

File: src/Language/PureScript/Errors.hs (hunk 2 above)
Comment (michaelficarra) [inline src/Language/PureScript/Errors.hs]: It's possible it would be helpful to show that.

Comment (garyb) [inline src/Language/PureScript/Errors.hs]: Something like this then?
```
Class import from ModuleName uses deprecated syntax that omits the 'class' keyword:

   ClassName

Should instead use the form:

  class ClassName
```

Likewise for exports:

```
Class export uses deprecated syntax that omits the 'class' keyword:

   ClassName

Should instead use the form:

  class ClassName
```

Comment (garyb) [inline src/Language/PureScript/Errors.hs]: I'll go ahead and make this change, it's certainly no worse than my original error message.

**Comment by garyb** [comment]
Updated with the new warnings, also updated the deprecated use of `qualified` warning to follow the same style.

**Comment by paf31** [comment]
:+1: Looks great, thanks!

**Comment by michaelficarra** [comment]
Someone's going to have to update the [language guide](https://github.com/purescript/purescript/wiki/Language-Guide#importing-modules).

**Comment by garyb** [comment]
Yeah, there will need to be a whole bunch of updating soon.

- deprecated `qualified` keyword
- deprecated imports without `class`
- type imports with no constructors no longer require `()`
- `import X` warns without qualification, an explicit list, or `(..)`
- operator aliases and correspondingly deprecated naming any function as an operator

I think we're going to 0.8 next most likely, as we're almost done with the issues in that milestone now.

**Pull request merged**